storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3MultiPartOutputStream.java

```diff
@@ -86,54 +86,73 @@ public void write(final byte[] b, final int off, final int len) throws IOExcepti
         if (b.length == 0) {
             return;
         }
-        final ByteBuffer source = ByteBuffer.wrap(b, off, len);
-        while (source.hasRemaining()) {
-            final int transferred = Math.min(partBuffer.remaining(), source.remaining());
-            final int offset = source.arrayOffset() + source.position();
-            // TODO: get rid of this array copying
-            partBuffer.put(source.array(), offset, transferred);
-            source.position(source.position() + transferred);
-            if (!partBuffer.hasRemaining()) {
-                flushBuffer(0, partSize);
+        try {
+            final ByteBuffer source = ByteBuffer.wrap(b, off, len);
+            while (source.hasRemaining()) {
+                final int transferred = Math.min(partBuffer.remaining(), source.remaining());
+                final int offset = source.arrayOffset() + source.position();
+                // TODO: get rid of this array copying
+                partBuffer.put(source.array(), offset, transferred);
+                source.position(source.position() + transferred);
+                if (!partBuffer.hasRemaining()) {
+                    flushBuffer(0, partSize);
+                }
             }
+        } catch (final Exception e) {
+            log.error("Failed to write to stream on upload {}, aborting transaction", uploadId, e);
+            abortUpload();
+            throw new IOException(e);
```

Comment:
Why IOException? And don't see this covered by tests

Reply:
I'm piggybacking on the existing checked exception. Also, first two tests are checking this type already when failing on writing or closing stream.
Let me know if I'm missing something.